### PR TITLE
Blog/custom service adapter

### PR DESCRIPTION
**Pull request by Skejven**
Feel free to fix typos and mistakes.
Should be merged together with https://github.com/Knotx/knotx-tutorials/pull/2

### CODE REVIEW

File: src/render/blog/custom-service-adapter.html.md [hunk 1]
```diff
@@ -0,0 +1,507 @@
...
+
+Now we have all the parts ready and can run the demo.
+Application directory should contain now the following artifacts:
+
+```
+├── knotx-standalone-1.0.0.json
+├── knotx-standalone-1.0.0.logback.xml
+├── app
+│   ├── knotx-standalone-1.0.0.fat.jar
+│   ├── knotx-standalone-1.0.0.fat.jar
```
Comment (toniedzwiedz) [inline src/render/blog/custom-service-adapter.html.md]: This line is a duplicate of the previous one. Should it be removed or changed?

**Comment by toniedzwiedz** [comment]
Alright, let's do it :)

**Comment by toniedzwiedz** [comment]
@Skejven I'm done. Please check if you're fine with the changes. We can also think of a more catchy title :)

File: src/render/blog/custom-service-adapter.html.md [hunk 2]
```diff
@@ -0,0 +1,507 @@
+---
+title: Custom Adapter Service
```
Comment (Skejven) [inline src/render/blog/custom-service-adapter.html.md]: You are right.
How about `Adapt Service without WebAPI`?

**Comment by Skejven** [comment]
Thank you for the comments and fixes. I've renamed tutorial to `Adapt Service without WebAPI`. Also tutorial codebase was updated.

**Comment by toniedzwiedz** [comment]
@Skejven I've been wondering, why spell _Web API_ like a single word? Do you actually mean the [Mozilla Standard](https://wiki.mozilla.org/WebAPI)? If not, I think it should be _Web API_ (a generic API served over HTTP), not _WebAPI_, which is very specific:

> _WebAPI_ is a term used to refer to a suite of device compatibility and access APIs that allow Web apps and content to access device hardware (such as battery status or the device vibration hardware), as well as access to data stored on the device (such as the calendar or contacts list). By adding these APIs, we hope to expand what the Web can do today and only proprietary platforms were able to do in the past.

File: src/render/blog/adapt-service-without-webapi.html.md [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+  public void stop() throws Exception {
+    // unregister adapter when no longer needed
+  }
+
+}
+```
+
+## Configuration
+
+Now we will create a simple configuration for our custom code. The configuration file defines a _Verticle_ that
```
Comment (marcinczeczko) [inline src/render/blog/adapt-service-without-webapi.html.md]: ...for our custom adapter.

File: src/render/blog/adapt-service-without-webapi.html.md [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+will initialise the whole _Service Adapter_ and enable us to pass properties to the custom class.
+
+Create `io.knotx.example.BooksDbAdapter.json` in `/src/main/resources/`:
+
+```json
+{
+  "main": "io.knotx.tutorials.BooksDbAdapter",
+  "options": {
+    "config": {
+      "address": "knotx.adapter.service.custom",
```
Comment (marcinczeczko) [inline src/render/blog/adapt-service-without-webapi.html.md]: make more meaningfull address name, e.g.: `knotx.adapter.service.booksdb`

File: src/render/blog/adapt-service-without-webapi.html.md [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+  public void init(Vertx vertx, Context context) {
+    super.init(vertx, context);
+    // using config() method from AbstractVerticle we simply pass our JSON file configuration to Java model
+    configuration = new BooksDbAdapterConfiguration(config());
+  }
+
+  @Override
+  public void start() throws Exception {
+    //we create RX version of vertx instance to easier handle reactive programming
+    final io.vertx.rxjava.core.Vertx rxVertx = new io.vertx.rxjava.core.Vertx(this.vertx);
```
Comment (marcinczeczko) [inline src/render/blog/adapt-service-without-webapi.html.md]: do we need fully qualified package name ? if we used AbstractVerticle class from rxjava package then this.vertx will give use rxified vertx object.

File: src/render/blog/adapt-service-without-webapi.html.md [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+    return adapterResponse;
+  }
+  
+}
+```
+
+What we do here is:
+- When there is a request in `processRequest`, the first thing we do is to get the `query` from the request object.
+- Then we create an [`Observable`](http://reactivex.io/documentation/observable.html) from the previously configured JDBC Client,
+ which gives us a `SQLConnection` object that will be used to perform the next operation.
```
Comment (marcinczeczko) [inline src/render/blog/adapt-service-without-webapi.html.md]: ... the next operation asynchoronously.

File: src/render/blog/adapt-service-without-webapi.html.md [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+    "knotx:io.knotx.example.BooksDbAdapter"
+  ],
+  "config": {
+    "knotx:io.knotx.ServiceKnot": {
+      "options": {
+        "config": {
+          "services": [
+            {
+              "name": "books-listing",
+              "address": "knotx.adapter.service.custom",
```
Comment (marcinczeczko) [inline src/render/blog/adapt-service-without-webapi.html.md]: update the address as i mentioned at the begining

File: src/render/blog/adapt-service-without-webapi.html.md [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+│   ├── local
+│       ├── books.html
+```
+
+You can run the _Knot.x_ instance using the following command:
+
+`java -Dlogback.configurationFile=knotx-standalone-1.0.0.logback.xml -cp "app/*" io.knotx.launcher.LogbackLauncher -conf knotx-standalone-1.0.0.json`
+
+When you visit the page [http://localhost:8092/content/local/books.html](http://localhost:8092/content/local/books.html),
+ you will see books and authors from the database listed.
```
Comment (marcinczeczko) [inline src/render/blog/adapt-service-without-webapi.html.md]: You can mention here some options to play. E.g. add new entries to the DB and refresh the page, so new books or authors will be visible immediately.

File: src/render/blog/adapt-service-without-webapi.html.md (hunk 5 above)
Comment (Skejven) [inline src/render/blog/adapt-service-without-webapi.html.md]: Right, but few lines below we register service with vertx `ProxyHelper` which requires plain Vertx `vertx`. So te solution you propose may look like in `start()` method:

```java
  @Override
  public void start() throws Exception {
    LOGGER.info("Starting <{}>", this.getClass().getSimpleName());

    final JDBCClient client = JDBCClient.createShared(this.vertx, configuration.getClientOptions());

    //register the service proxy on event bus
    consumer = ProxyHelper
        .registerService(AdapterProxy.class, this.getVertx(),
            new BooksDbAdapterProxyImpl(client),
            configuration.getAddress());
  }
```

I think it is much more confusing. Using `this.vertx` to get rxfied version just to use `this.getVertx()` to get Vertx version is very unintuitive. What do you think?

File: src/render/blog/adapt-service-without-webapi.html.md (hunk 8 above)
Comment (Skejven) [inline src/render/blog/adapt-service-without-webapi.html.md]: Done

**Pull request merged**